src/NanopassSharp/Pass.cs

```diff
@@ -0,0 +1,71 @@
+﻿using System.Collections.Generic;
+
+namespace NanopassSharp;
+
+/// <summary>
+/// A single compiler pass.
+/// </summary>
+/// <param name="Name">The name of the pass.</param>
+/// <param name="Documentation">The pass' corresponding documentation.</param>
+/// <param name="Transformations">The transformations the pass applies to its nodes.</param>
+/// <param name="Tree">The original node tree of the pass.</param>
+public sealed record class Pass
+(
+    string Name,
+    string? Documentation,
+    Pass Previous,
+    IList<ITransformation> Transformations,
+    // Should this be immutable?
+    // Should the tree be the tree resulting from applying the pass' transformations?
+    // Should it be created immediately by the language-specific parser
+    // or be based on the previous pass?
+    TypeNodeTree Tree
+);
+
+/// <summary>
+/// A tree of nodes representing types which are the input and output of a pass.
+/// </summary>
+/// <param name="Root">The root node.</param>
+/// <param name="Nodes">The nodes of the tree.</param>
+public sealed record class TypeNodeTree
+(
+    TypeNode Root,
```

Comment:
Is there a possibility for multiple root nodes? Or is this a non-concern for now?

Reply:
Depends on what kind of hierarchy we're trying to model, I suppose. In languages like C# and Java, nested types exist which would naturally lead to tree-like structure, but in other langs, the types would just have to be laid out flat and the tree structure implied by the referenced types in each type.

Reply:
I'd say that you are still modeling a hierarchy, what this compiles to is a detail.